File: src/pwSquaredDist.c
/*
 * pwSquaredDist.c
 * copyright (c) by Alexander Lehmann <afwlehmann@googlemail.com>
 */

#include <R.h>
#include <Rdefines.h>

/**
 * Squared distance of each pair of row vectors of the given
 * matrix `X` with dimensions `rows` x `cols`. The result is
 * stored in `d` which must be a vector of length N*(N-1)/2
 * where N = `rows`.
 */
void pwSquaredDist(double *X, int *rows, int *cols, double *d) {
    int i, k;
    double *dd, m, aux;
    register int base, j;

    base = 0;
    for (k = 0; k < *cols; ++k) {
        dd = d;
        for (i = 0; i < *rows; ++i) {
            m = X[base + i];
            for (j = i+1; j < *rows; ++j) {
                aux = X[base + j] - m;
                *dd++ += aux * aux;
            }
        }
        base += *rows;
    }
}

File: R/pwSquaredDist.R
#
# pwSquaredDist.R
# copyright (c) by Alexander Lehmann <afwlehmann@googlemail.com>
#

#' Squared distance of each pair of row vectors of the matrix \code{X}.
#'
#' @examples
#' Y <- matrix(c(1, -2, 3, 55, -8, 4, 3, 7, 11), ncol=3)
#' distances <- pwSquaredDist(Y)        # Result as vector
#'
#' D <- pwSquaredDist(Y, asMatrix=TRUE) # Result as lower triangular matrix
#' @param X a matrix
#' @param asMatrix a boolean indicating whether the result should be a
#'   vector or a symmetric matrix, defaults to \code{FALSE}
#' @return a vector of length N*(N-1)/2 if \code{asMatrix=T}, else a NxN
#'   symmetric matrix
#' @export
pwSquaredDist <- function(X, asMatrix=FALSE) {
  N <- nrow(X)
  M <- ncol(X)
  d <- .C("pwSquaredDist",
          as.double(X), N, M,
          d = double(N * (N-1) / 2),
          PACKAGE="rtools")$d
  if (asMatrix) {
    mat <- matrix(0, N, N)
    mat[lower.tri(mat)] <- d
    mat + t(mat)
  } else
    d
}
